File: java/quantile-estimator/src/test/java/org/mydotey/quantile/timewindow/TimeWindowQuantileEstimatorTest.java
package org.mydotey.quantile.timewindow;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.function.BiFunction;
import java.util.function.Supplier;

import org.junit.Assert;
import org.mydotey.quantile.QuantileEstimator;
import org.mydotey.quantile.QuantileEstimatorTest;
import org.mydotey.quantile.facade.QuantileEstimators;
import org.mydotey.quantile.value.IntCalculator;
import org.mydotey.quantile.value.IntComparator;
import org.mydotey.quantile.value.LongCalculator;
import org.mydotey.quantile.value.LongComparator;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
public class TimeWindowQuantileEstimatorTest extends QuantileEstimatorTest {

    @Override
    protected void addQuantileEsimators() {
        Supplier<QuantileEstimator<Long>> longSupplier = () -> QuantileEstimators
                .newClassicEstimator(LongComparator.DEFAULT, LongCalculator.DEFAULT);
        _estimators.put(Long.class, QuantileEstimators.newTimeWindowEstimator(longSupplier,
                TimeUnit.SECONDS.toMillis(5), TimeUnit.SECONDS.toMillis(1)));

        Supplier<QuantileEstimator<Integer>> intSupplier = () -> QuantileEstimators
                .newClassicEstimator(IntComparator.DEFAULT, IntCalculator.DEFAULT);
        _estimators.put(Integer.class, QuantileEstimators.newTimeWindowEstimator(intSupplier,
                TimeUnit.SECONDS.toMillis(5), TimeUnit.SECONDS.toMillis(1)));
    }

    @Override
    protected void test(QuantileEstimator<Integer> quantileEstimator, List<Double> quantiles, int count, int upperBound,
            double errorRate, BiFunction<Integer, Integer, List<Integer>> dataProvider) {
        int maxError = (int) (upperBound * errorRate);
        List<Integer> items = dataProvider.apply(count, upperBound);

        int timeWindowSec = 10;
        for (int i = 0; i < timeWindowSec; i++) {
            items.forEach(item -> quantileEstimator.add(item));
            try {
                Thread.sleep(TimeUnit.SECONDS.toMillis(1));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        System.out.println("data: " + items);
        System.out.println();

        Collections.sort(items);
        System.out.println("sorted: " + items);
        System.out.println();

        HashMap<Double, Integer> quantileResults = new HashMap<>();
        for (Double quantile : quantiles) {
            int pos = (int) (count * quantile) - 1;
            if (pos < 0)
                pos = 0;

            int item = items.get(pos);
            quantileEstimator.add(item);
            quantileResults.put(quantile, item);
        }

        Map<Double, Integer> results = quantileEstimator.get(quantiles);
        for (int i = 0; i < quantiles.size(); i++) {
            Double quantile = quantiles.get(i);
            Integer expected = quantileResults.get(quantile);
            Integer actual = results.get(quantile);
            Integer actualError = Math.abs(actual - expected);
            System.out.println("quantile " + quantile + ", expected: " + expected + ", actual: " + actual + ", error: "
                    + actualError);
            System.out.println();

            Assert.assertTrue("error > " + maxError, actualError <= maxError);
        }
    }

}

File: dotnet/test/MyDotey.QuantileEstimator.Tests/TimeWindow/TimeWindowQuantileEstimatorTest.cs
using System;
using System.Collections.Generic;
using System.Threading;

using MyDotey.Quantile.Value;
using MyDotey.Quantile.Facade;
using Xunit;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
namespace MyDotey.Quantile.Tests
{
    public class TimeWindowQuantileEstimatorTest : QuantileEstimatorTest
    {
        protected override void AddQuantileEsimators()
        {
            Func<IQuantileEstimator<long>> longSupplier = () => QuantileEstimators
                    .NewClassicEstimator(LongComparer.DEFAULT, LongCalculator.DEFAULT);
            _estimators[typeof(long)] = QuantileEstimators.NewTimeWindowEstimator(longSupplier,
                    5 * 1000, 1 * 1000);

            Func<IQuantileEstimator<int>> intSupplier = () => QuantileEstimators
                    .NewClassicEstimator(IntComparer.DEFAULT, IntCalculator.DEFAULT);
            _estimators[typeof(int)] = QuantileEstimators.NewTimeWindowEstimator(intSupplier,
                    5 * 1000, 1 * 1000);
        }

        protected override void Test(IQuantileEstimator<int> quantileEstimator, List<Double> quantiles, int count, int upperBound,
                double errorRate, Func<int, int, List<int>> dataProvider)
        {
            int maxError = (int)(upperBound * errorRate);
            List<int> items = dataProvider(count, upperBound);

            int timeWindowSec = 10;
            for (int i = 0; i < timeWindowSec; i++)
            {
                items.ForEach(item => quantileEstimator.Add(item));
                Thread.Sleep(1 * 1000);
            }

            Console.WriteLine("data: " + items);
            Console.WriteLine();

            items.Sort();

            Console.WriteLine("sorted: " + items);
            Console.WriteLine();

            Dictionary<Double, int> quantileResults = new Dictionary<Double, int>();
            foreach (Double quantile in quantiles)
            {
                int pos = (int)(count * quantile) - 1;
                if (pos < 0)
                    pos = 0;

                int item = items[pos];
                quantileEstimator.Add(item);
                quantileResults[quantile] = item;
            }

            Dictionary<Double, int> results = quantileEstimator.Get(quantiles);
            for (int i = 0; i < quantiles.Count; i++)
            {
                Double quantile = quantiles[i];
                int expected = quantileResults[quantile];
                int actual = results[quantile];
                int actualError = Math.Abs(actual - expected);
                Console.WriteLine("quantile " + quantile + ", expected: " + expected + ", actual: " + actual + ", error: "
                        + actualError);
                Console.WriteLine();

                Assert.True(actualError <= maxError);
            }
        }
    }
}